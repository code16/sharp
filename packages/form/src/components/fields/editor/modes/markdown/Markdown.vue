<template>
    <div>
        <LocalizedEditors
            :editor="editor"
            :name="uniqueIdentifier"
            :value="value"
            :locale="locale"
            :locales="locales"
            :create-editor="createEditor"
            v-slot="{ editor, locale }"
        >
            <SharpEditor
                :editor="editor"
                :toolbar-options="toolbarOptions(editor)"
                v-bind="$props"
                @update="handleUpdate({ editor, locale, ...$event })"
            />
        </LocalizedEditors>
    </div>
</template>

<script>
    import { Markdown } from 'tiptap-markdown';
    import { Editor } from '@tiptap/vue-2';
    import { lang } from "sharp";
    import SharpEditor from '../../Editor';
    import { defaultEditorOptions, editorProps } from "../..";
    import { LocalizedEditor } from '../../../../../mixins/localize/editor';
    import LocalizedEditors from "../../LocalizedEditors";
    import { normalizeText } from "../../../../../util/text";


    export default {
        mixins: [
            LocalizedEditor
        ],
        components: {
            SharpEditor,
            LocalizedEditors,
        },
        props: {
            ...editorProps,
            extensions: Array,
            toolbar: Array,
            nl2br: Boolean,
            tightListsOnly: Boolean,

            readOnly: Boolean,
            uniqueIdentifier: String,
        },
        data() {
            return {
                editor: null,
            }
        },
        methods: {
            handleUpdate({ editor, locale, error }) {
                const content = normalizeText(editor.storage.markdown.getMarkdown() ?? '');
                this.$emit('input', this.localizedValue(content, locale), { error });
            },
            toolbarOptions(editor) {
                const options = [];
                const hasList = this.toolbar?.some(button => button === 'bullet-list' || button === 'ordered-list');

                if(!this.tightListsOnly && hasList) {
                    options.push({
                        command: () => editor.chain().toggleTight().run(),
                        disabled: !editor.can().toggleTight(),
                        label: lang('form.editor.dropdown.options.toggle_tight_list'),
                    });
                }

                return options;
            },

            createEditor({ content }) {
                return new Editor({
                    ...defaultEditorOptions,
                    extensions: [
                        ...this.extensions,
                        Markdown.configure({
                            breaks: this.nl2br,
                        }),
                    ],
                    content,
                    editable: !this.readOnly,
                });
            },
        },
        created() {
            if(!this.isLocalized) {
                this.editor = this.createEditor({
                    content: this.localizedText,
                });
            }
        },
        beforeDestroy() {
            this.editor?.destroy();
        },
    }
</script>
