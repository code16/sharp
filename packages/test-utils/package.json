{
    "name": "@sharp/test-utils",
    "private": true,
    "main": "src/index.js",
    "version": "1.0.0"
}