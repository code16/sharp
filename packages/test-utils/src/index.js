export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC, unmockSFC } from './mockSFC';
export { mockProperty, unmockProperty, setter } from './mock-utils';
export { nextRequestFulfilled } from './moxios-utils';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}