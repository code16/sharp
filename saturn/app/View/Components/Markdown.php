<?php

namespace App\View\Components;

use Illuminate\Contracts\View\View;
use Illuminate\View\Component;
use League\CommonMark\GithubFlavoredMarkdownConverter;
use League\CommonMark\MarkdownConverterInterface;

class Markdown extends Component
{
    protected array $options;

    public function __construct(array $options = [])
    {
        $this->options = $options;
    }

    public function render(): View
    {
        return view('components.markdown');
    }

    public function toHtml(string $markdown): string
    {
        return $this->converter()->convertToHtml($markdown);
    }

    protected function converter(): MarkdownConverterInterface
    {
        $options = array_merge($this->options, [
            'html_input' => 'allow',
            'renderer'   => [
                'soft_break' => config('sharp.markdown_editor.nl2br') ? '<br>' : "\n",
            ],
        ]);

        return new GithubFlavoredMarkdownConverter($options);
    }
}
