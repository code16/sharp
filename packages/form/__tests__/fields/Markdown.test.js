import Vue from 'vue';
import Markdown from '../../src/components/fields/markdown/Markdown.vue';

import { MockI18n, MockInjections } from '@sharp/test-utils';
import { mount } from '@vue/test-utils';

import SimpleMDE from 'simplemde';


jest.mock('../../src/components/fields/upload/VueClip', ()=>({
    data: ()=>({ uploader:({ _uploader:{ hiddenFileInput:{ click:jest.fn() } } }) }),
    render: h=>h()
}));

describe('markdown-field', () => {
    Vue.use(MockI18n);

    async function createWrapper(customOptions={}) {
        let { propsData, ...options } = customOptions;
        const wrapper = mount(Markdown, {
            attachToDocument: true,
            provide: MockInjections.provide,
            propsData: {
                value: { text:'' },
                readOnly: false,
                placeholder: 'Champ md',
                height: 310,
                innerComponents: { upload:{ maxImageSize:3 } },
                fieldConfigIdentifier: 'my_markdown',
                uniqueIdentifier: 'my_markdown',
                toolbar: ['bold', 'italic'],
                ...propsData
            },
            ...options
        });
        await wrapper.vm.$nextTick();
        return wrapper;
    }

    function createLocalizedWrapper({ value, locale, locales }) {
        return createWrapper({
            propsData: { value, locale },
            computed: {
                isLocalized:()=>true,
                locales:()=>locales
            }
        });
    }

    beforeAll(()=>{
        // prevent jsdom errors
        window.focus = jest.fn();
        window.scrollTo = jest.fn();
    });

    beforeEach(()=>{
        // mock range functions
        document.body.createTextRange = () => ({
            getBoundingClientRect: () => ({ }),
            getClientRects: () => ({ })
        });

        MockI18n.mockLangFunction();
    });

    afterEach(() => {
        jest.resetAllMocks();
    });

    describe('basic tests', () => {
        test('can mount Markdown field', async () => {
            let wrapper = await createWrapper();

            expect(wrapper.html()).toMatchSnapshot();
        });

        test('can mount "localized" Markdown field', async () => {
            let wrapper = await createLocalizedWrapper({
                value: { text:{ fr:'', en: '' } },
                locales: ['fr', 'en'],
                locale: 'fr'
            });

            expect(wrapper.html()).toMatchSnapshot();
        });


        test('can mount "read only" Markdown field', async () => {
           let wrapper = await createWrapper({
                propsData: {
                    readOnly: true
                }
            });

            expect(wrapper.html()).toMatchSnapshot();
        });

        test('handle locale changed', async () => {
            let wrapper = await createLocalizedWrapper({
                value: { text:{ fr:'', en: '' } },
                locales:['fr', 'en'],
                locale: 'fr',
            });

            wrapper.setProps({ locale:'en' });
            wrapper.setMethods({ refreshOnExternalChange: jest.fn() });

            await wrapper.vm.$nextTick();

            expect(wrapper.vm.refreshOnExternalChange).toHaveBeenCalled();
        });

        test('localized: simplemde instances, current simplemde', async () => {
            let wrapper = await createLocalizedWrapper({
                value: { text:{ fr:'', en: '' } },
                locales:['fr', 'en'],
                locale: 'fr',
            });

            expect(wrapper.vm.simplemdeInstances).toEqual({ 'fr':expect.any(SimpleMDE), 'en':expect.any(SimpleMDE) });
            expect(wrapper.vm.simplemde).toBe(wrapper.vm.simplemdeInstances['fr']);
            expect(wrapper.vm.codemirror).toBe(wrapper.vm.simplemdeInstances['fr'].codemirror);

            wrapper.setProps({ locale: 'en' });

            expect(wrapper.vm.simplemde).toBe(wrapper.vm.simplemdeInstances['en']);
            expect(wrapper.vm.codemirror).toBe(wrapper.vm.simplemdeInstances['en'].codemirror);
        });

        test('simplemde instances, current simplemde', async () => {
            let wrapper = await createWrapper();
            expect(wrapper.vm.simplemdeInstances).toEqual(expect.any(SimpleMDE));
            expect(wrapper.vm.simplemde).toEqual(wrapper.vm.simplemdeInstances);
            expect(wrapper.vm.codemirror).toEqual(wrapper.vm.simplemdeInstances.codemirror);
        });

        test('filesByName', async () => {
            let wrapper = await createWrapper({
                propsData: {
                    value: {
                        files: [{ name:'aaa.jpg' }, { name:'bbb.jpg' }]
                    }
                }
            });
            expect(wrapper.vm.filesByName).toMatchObject({
                'aaa.jpg': { name:'aaa.jpg' },
                'bbb.jpg': { name:'bbb.jpg' }
            });
        });

        test('indexByFileId', async () => {
            let wrapper = await createWrapper();
            let id = wrapper.vm.idSymbol;
            wrapper.setProps({
                value: {
                    files: [{ [id]:1 }, { [id]:6 }]
                }
            });
            expect(wrapper.vm.indexByFileId).toEqual({
                1: 0,
                6: 1
            });
        });

        test('createSimpleMDE', async () => {
            let wrapper = await createWrapper({
                propsData:{
                    value: { text: 'value' },
                    toolbar: ['bold']
                },
                created() {
                    jest.spyOn(this, 'createSimpleMDE');
                }
            });
            let { element } = wrapper.find({ ref:'textarea' });
            expect(wrapper.vm.createSimpleMDE).toHaveBeenCalledWith({ element, initialValue:'value' });
            expect(wrapper.vm.simplemdeInstances.options).toMatchObject({
                initialValue: 'value',
                placeholder: 'Champ md',
                spellChecker: false,
                autoDownloadFontAwesome: false,
                toolbar: [{
                    action: expect.any(Function),
                    className: 'fas fa-bold',
                    name: 'bold',
                    title: '{{ form.markdown.icons.bold.title }}'
                }],
            });
        });

        test('localizedTextareaRef', async () => {
            let wrapper = await createWrapper();
            expect(wrapper.vm.localizedTextareaRef('fr')).toEqual('textarea_fr');
        });

        test('bound toolbar buttons custom action properly', async () =>{
            let wrapper = await createWrapper({
                propsData: {
                    toolbar: ['image']
                }
            });

            let { simplemde } = wrapper.vm;
            wrapper.vm.insertUploadImage = jest.fn();

            expect(simplemde.toolbar[0].action).toBeInstanceOf(Function);
            simplemde.toolbar[0].action();

            expect(wrapper.vm.insertUploadImage).toHaveBeenCalled();
        });

        test('set read only properly', async () => {
            let wrapper = await createWrapper();

            let { simplemde, codemirror } = wrapper.vm;

            expect(codemirror.getOption('readOnly')).toBe(false);

            wrapper.vm.setReadOnly(simplemde);

            expect(codemirror.getOption('readOnly')).toBe(true);
        });

        test('add codemirror event listener properly', async () => {
            let wrapper = await createWrapper();

            let { codemirror } = wrapper.vm;

            let callback = jest.fn();
            codemirror.on = jest.fn();

            wrapper.vm.codemirrorOn(codemirror, 'event', callback);
            expect(codemirror.on).toHaveBeenCalledWith('event', callback);
            expect(callback).not.toHaveBeenCalled();

            wrapper.vm.codemirrorOn(codemirror, 'event', callback, true);
            expect(codemirror.on).toHaveBeenCalledTimes(2);
            expect(codemirror.on).toHaveBeenLastCalledWith('event', callback);
            expect(callback).toHaveBeenCalled();
        });

        test('emit input on text changed', async () => {
            let wrapper = await createWrapper();

            let { simplemde } = wrapper.vm;

            wrapper.setMethods({
                localizedValue: jest.fn(()=>'localizedValue')
            });

            simplemde.value('AAA');

            expect(wrapper.emitted().input).toHaveLength(1);
            expect(wrapper.emitted().input[0]).toEqual(['localizedValue'])
        });

        test('mount with null localized value', async () => {
            let wrapper = await createLocalizedWrapper({ value: { text: null }, locales:['fr', 'en'], locale:'fr' });
            expect(wrapper.find(Markdown).isVueInstance()).toBe(true);
        });
    });


    describe('uploader insertion', () => {
        function triggerUploadFile(uploader) {
            setTimeout(() => uploader.$emit('added'), 0);
        }

        function mockCodemirror(codemirror, { onWidgetAdded }={}) {
            codemirror.markText = jest.fn((from,to, { $component }) => {
                if($component && onWidgetAdded) {
                    onWidgetAdded($component);
                }
                return {
                    on: jest.fn(),
                    clear: jest.fn(),
                    lines: [{ on: jest.fn() }],
                }
            });
            jest.spyOn(codemirror, 'setSelection');
            jest.spyOn(codemirror, 'replaceRange');
        }

        const mockMixin = {
            watch: {
                codemirror: {
                    sync: true,
                    handler(codemirror) {
                        mockCodemirror(codemirror, {
                            onWidgetAdded: $widgetComp => this._recordedWidgets.push($widgetComp)
                        });
                    }
                },
            },
            methods: {
                createUploader(...args) {
                    let $uploader = Markdown.methods.createUploader.apply(this, args);
                    triggerUploadFile($uploader);
                    return $uploader;
                }
            },
            created() {
                jest.spyOn(this, 'insertUploadImage');
                jest.spyOn(this, 'removeMarker');
                this._recordedWidgets = [];
            }
        };

        test('insert image uploader and text properly', async () => {
            let wrapper = await createWrapper({
                mixins: [mockMixin],
                propsData: {
                    value: { text: 'Lorem Elsass ipsum' }
                }
            });

            wrapper.vm.codemirror.setSelection({line: 0, ch: 5}, {line: 0, ch: 13});

            await wrapper.vm.insertUploadImage({ isInsertion: true });

            expect(wrapper.vm.simplemde.value()).toBe('Lorem\n![]()\nipsum');
        });

        test('update image uploader and text properly', async () => {
            let wrapper = await createWrapper({
                mixins:[mockMixin]
            });

            let $uploader = await wrapper.vm.insertUploadImage({ isInsertion:true });

            expect(wrapper.vm.simplemde.value()).toBe('\n![]()\n');

            $uploader.marker.find = jest.fn(() => ({ from:{ line: 1, ch: 0 }, to:{ line: 1, ch:5 }}) );
            $uploader.$emit('success', { name: 'cat.jpg' });

            expect(wrapper.vm.simplemde.value()).toBe('\n![](cat.jpg)\n');

            expect(wrapper.vm.value.files).toEqual([{
                [wrapper.vm.idSymbol]: 0,
                name: 'cat.jpg'
            }]);
        });

        test('parse and insert image uploader and text properly', async () => {
            let wrapper = await createWrapper({
                mixins: [mockMixin],
                propsData: {
                    value: {
                        text:'aaa\n![Cat](cat.jpg)\nbbb',
                        files:[{
                            name:'cat.jpg',
                            size: 123
                        }]
                    },
                },
            });
            /// Parse markdown content
            wrapper.vm.$tab.$emit('active');

            expect(wrapper.vm.simplemde.value()).toBe('aaa\n![Cat](cat.jpg)\nbbb');

            expect(wrapper.vm.value.files).toEqual([{
                [wrapper.vm.idSymbol]: 0,
                name:'cat.jpg',
                size: 123
            }]);

            expect(wrapper.vm.codemirror.setSelection).toHaveBeenCalledTimes(1);
            expect(wrapper.vm.codemirror.setSelection).toHaveBeenCalledWith({ line:1, ch:0 }, { line:1, ch:15 }, expect.anything());

            expect(wrapper.vm.insertUploadImage).toHaveBeenCalledTimes(1);
            expect(wrapper.vm.insertUploadImage).toHaveBeenCalledWith({ replaceBySelection:true, data:{ name:'cat.jpg', title:'Cat' } });

            expect(wrapper.vm._recordedWidgets).toHaveLength(1);
            expect(wrapper.vm._recordedWidgets[0].value).toMatchObject({
                name: 'cat.jpg',
                size: 123
            });
        });

        test('delete properly', async () => {
            let wrapper = await createWrapper({
                mixins: [mockMixin],
                propsData: {
                    value: {
                        text:'aaa\n![Cat](cat.jpg)\nbbb',
                        files:[{
                            name:'cat.jpg',
                            size: 123
                        }]
                    },
                }
            });
            /// Parse markdown content
            wrapper.vm.$tab.$emit('active');

            let $uploader = wrapper.vm._recordedWidgets[0];

            jest.spyOn($uploader, '$destroy');

            wrapper.vm.removeMarker($uploader, { isCMEvent: true });

            expect($uploader.$destroy).toHaveBeenCalled();
            expect(wrapper.vm.value.files).toEqual([]);
        });

        test('register delete event and delete properly on event', async () => {
            let wrapper = await createWrapper({
                mixins:[mockMixin]
            });
            let $uploader = await wrapper.vm.insertUploadImage({ isInsertion:true });

            const lineOn = $uploader.marker.lines[0].on;

            expect(lineOn).toHaveBeenCalledWith('delete', expect.any(Function));

            let deleteHandler = lineOn.mock.calls[0][1];

            deleteHandler();

            expect(wrapper.vm.removeMarker).toHaveBeenCalledTimes(1);
            expect(wrapper.vm.removeMarker).toHaveBeenCalledWith($uploader, { isCMEvent: true, relativeFallbackLine: 1 });
        });


        test('Delete properly removing from upload component', async () => {
            let wrapper = await createWrapper({
                mixins:[mockMixin]
            });

            let $uploader = await wrapper.vm.insertUploadImage({ isInsertion:true });

            $uploader.marker.find = jest.fn(()=>({from:{ line: 1, ch:0 }, to:{ line:1, ch:5 }}));

            jest.spyOn($uploader, '$destroy');

            $uploader.$emit('remove');

            expect(wrapper.vm.removeMarker).toHaveBeenCalledWith($uploader, { relativeFallbackLine: 1 });

            expect(wrapper.vm.codemirror.replaceRange).toHaveBeenCalledWith('', { line: 0 }, { line: 2, ch:0 });

            expect($uploader.$destroy).toHaveBeenCalled();
        });

        test('expose appropriate props to markdown upload component', async () =>{
            let wrapper = await createWrapper({
                mixins: [mockMixin],
                propsData: {
                    value: {
                        text:'aaa\n![Cat](cat.jpg)\nbbb',
                        files:[{
                            name:'cat.jpg',
                            size: 123
                        }]
                    },
                },
            });

            /// Parse markdown content
            wrapper.vm.$tab.$emit('active');

            expect(wrapper.vm._recordedWidgets[0].$props).toMatchObject({
                id: 0,
                value: {
                    name:'cat.jpg',
                    size: 123
                },
                maxImageSize: 3,
                downloadId: 'my_markdown',
                pendingKey: 'my_markdown.upload.0'
            });

            await wrapper.vm.insertUploadImage();

            expect(wrapper.vm._recordedWidgets[1].$props).toMatchObject({
                id: 1,
                value: undefined,
                maxImageSize: 3,
                downloadId: 'my_markdown',
                pendingKey: 'my_markdown.upload.1'
            });
        });

        test('index files correctly on mounted', async () => {
            let wrapper = await createWrapper({
                propsData: {
                    value: {
                        files:[{
                            name:'cat.jpg',
                            size: 123
                        }, {
                            name:'dog.png',
                            size: 456
                        }]
                    }
                }
            });

            expect(wrapper.vm.value.files).toEqual([{
                [wrapper.vm.idSymbol]: 0,
                name:'cat.jpg',
                size: 123
            }, {
                [wrapper.vm.idSymbol]: 1,
                name:'dog.png',
                size: 456
            }])
        });

        test('refresh properly', async () => {
            let wrapper = await createWrapper({ mixins:[mockMixin] });

            let { simplemde } = wrapper.vm;
            let { codemirror } = simplemde;

            let $uploader = await wrapper.vm.insertUploadImage();

            codemirror.refresh = jest.fn();
            codemirror.focus = jest.fn();

            $uploader.$emit('refresh');

            expect(codemirror.refresh).toHaveBeenCalled();
            expect(codemirror.focus).toHaveBeenCalled();
        });

        test('update file data properly', async () => {
            let wrapper = await createWrapper({ mixins:[mockMixin] });
            let $uploader = await wrapper.vm.insertUploadImage({ isInsertion: true });

            $uploader.marker.find = jest.fn(() => ({ from:{ line: 1, ch: 0 }, to:{ line: 1, ch:5 }}) );

            $uploader.$emit('success', { name: 'cat.jpg' });
            $uploader.$emit('update', { cropData: { } });

            expect(wrapper.vm.value.files).toMatchObject([{
                name: 'cat.jpg',
                cropData: { }
            }]);
        })

    });
});

