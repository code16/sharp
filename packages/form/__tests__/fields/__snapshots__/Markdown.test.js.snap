// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown-field basic tests can mount "localized" Markdown field 1`] = `
<div class="SharpMarkdown">
	<div class="SharpModule__inner">
		<div>
			<textarea id="textarea_fr" style="display: none;"></textarea>
			<div class="editor-toolbar">
				<a title="{{ form.markdown.icons.bold.title }} (Ctrl-B)" tabindex="-1" class="fas fa-bold"></a>
				<a title="{{ form.markdown.icons.italic.title }} (Ctrl-I)" tabindex="-1" class="fas fa-italic"></a>
			</div>
			<div class="CodeMirror cm-s-paper CodeMirror-wrap CodeMirror-empty" style="width: auto; height: 310px;">
				<div style="overflow: hidden; position: relative; width: 3px; height: 0px;">
					<textarea style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;" autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0"></textarea>
				</div>
				<div class="CodeMirror-vscrollbar" cm-not-content="true">
					<div style="min-width: 1px;"></div>
				</div>
				<div class="CodeMirror-hscrollbar" cm-not-content="true">
					<div style="height: 100%; min-height: 1px;"></div>
				</div>
				<div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div>
				<div class="CodeMirror-gutter-filler" cm-not-content="true"></div>
				<div class="CodeMirror-scroll" tabindex="-1">
					<div class="CodeMirror-sizer" style="margin-left: 0px;">
						<div style="position: relative;">
							<div class="CodeMirror-lines" role="presentation">
								<div style="position: relative; outline: none;" role="presentation">
									<pre style="height: 0px; overflow: visible;" class="CodeMirror-placeholder">Champ md</pre>
									<div class="CodeMirror-measure">
										<pre><span>xxxxxxxxxx</span></pre>
									</div>
									<div class="CodeMirror-measure"></div>
									<div style="position: relative; z-index: 1;"></div>
									<div class="CodeMirror-cursors"></div>
									<div class="CodeMirror-code" role="presentation"></div>
								</div>
							</div>
						</div>
					</div>
					<div style="position: absolute; height: 30px; width: 1px;"></div>
					<div class="CodeMirror-gutters" style="display: none;"></div>
				</div>
			</div>
			<div class="editor-preview-side"></div>
		</div>
		<div style="display: none;">
			<textarea id="textarea_en" style="display: none;"></textarea>
			<div class="editor-toolbar">
				<a title="{{ form.markdown.icons.bold.title }} (Ctrl-B)" tabindex="-1" class="fas fa-bold"></a>
				<a title="{{ form.markdown.icons.italic.title }} (Ctrl-I)" tabindex="-1" class="fas fa-italic"></a>
			</div>
			<div class="CodeMirror cm-s-paper CodeMirror-wrap CodeMirror-empty" style="width: auto; height: 310px;">
				<div style="overflow: hidden; position: relative; width: 3px; height: 0px;">
					<textarea style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;" autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0"></textarea>
				</div>
				<div class="CodeMirror-vscrollbar" cm-not-content="true">
					<div style="min-width: 1px;"></div>
				</div>
				<div class="CodeMirror-hscrollbar" cm-not-content="true">
					<div style="height: 100%; min-height: 1px;"></div>
				</div>
				<div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div>
				<div class="CodeMirror-gutter-filler" cm-not-content="true"></div>
				<div class="CodeMirror-scroll" tabindex="-1">
					<div class="CodeMirror-sizer" style="margin-left: 0px;">
						<div style="position: relative;">
							<div class="CodeMirror-lines" role="presentation">
								<div style="position: relative; outline: none;" role="presentation">
									<pre style="height: 0px; overflow: visible;" class="CodeMirror-placeholder">Champ md</pre>
									<div class="CodeMirror-measure">
										<pre><span>xxxxxxxxxx</span></pre>
									</div>
									<div class="CodeMirror-measure"></div>
									<div style="position: relative; z-index: 1;"></div>
									<div class="CodeMirror-cursors"></div>
									<div class="CodeMirror-code" role="presentation"></div>
								</div>
							</div>
						</div>
					</div>
					<div style="position: absolute; height: 30px; width: 1px;"></div>
					<div class="CodeMirror-gutters" style="display: none;"></div>
				</div>
			</div>
			<div class="editor-preview-side"></div>
		</div>
	</div>
</div>
`;

exports[`markdown-field basic tests can mount "read only" Markdown field 1`] = `
<div class="SharpMarkdown SharpMarkdown--read-only">
	<div class="SharpModule__inner">
		<textarea style="display: none;"></textarea>
		<div class="editor-toolbar">
			<a title="{{ form.markdown.icons.bold.title }}" tabindex="-1" class="fas fa-bold"></a>
			<a title="{{ form.markdown.icons.italic.title }}" tabindex="-1" class="fas fa-italic"></a>
		</div>
		<div class="CodeMirror cm-s-paper CodeMirror-wrap CodeMirror-empty" style="width: auto; height: 310px;">
			<div style="overflow: hidden; position: relative; width: 3px; height: 0px;">
				<textarea style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;" autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0"></textarea>
			</div>
			<div class="CodeMirror-vscrollbar" cm-not-content="true">
				<div style="min-width: 1px;"></div>
			</div>
			<div class="CodeMirror-hscrollbar" cm-not-content="true">
				<div style="height: 100%; min-height: 1px;"></div>
			</div>
			<div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div>
			<div class="CodeMirror-gutter-filler" cm-not-content="true"></div>
			<div class="CodeMirror-scroll" tabindex="-1">
				<div class="CodeMirror-sizer" style="margin-left: 0px;">
					<div style="position: relative;">
						<div class="CodeMirror-lines" role="presentation">
							<div style="position: relative; outline: none;" role="presentation">
								<pre style="height: 0px; overflow: visible;" class="CodeMirror-placeholder">Champ md</pre>
								<div class="CodeMirror-measure">
									<pre><span>xxxxxxxxxx</span></pre>
								</div>
								<div class="CodeMirror-measure"></div>
								<div style="position: relative; z-index: 1;"></div>
								<div class="CodeMirror-cursors"></div>
								<div class="CodeMirror-code" role="presentation"></div>
							</div>
						</div>
					</div>
				</div>
				<div style="position: absolute; height: 30px; width: 1px;"></div>
				<div class="CodeMirror-gutters" style="display: none;"></div>
			</div>
		</div>
		<div class="editor-preview-side"></div>
	</div>
</div>
`;

exports[`markdown-field basic tests can mount Markdown field 1`] = `
<div class="SharpMarkdown">
	<div class="SharpModule__inner">
		<textarea style="display: none;"></textarea>
		<div class="editor-toolbar">
			<a title="{{ form.markdown.icons.bold.title }} (Ctrl-B)" tabindex="-1" class="fas fa-bold"></a>
			<a title="{{ form.markdown.icons.italic.title }} (Ctrl-I)" tabindex="-1" class="fas fa-italic"></a>
		</div>
		<div class="CodeMirror cm-s-paper CodeMirror-wrap CodeMirror-empty" style="width: auto; height: 310px;">
			<div style="overflow: hidden; position: relative; width: 3px; height: 0px;">
				<textarea style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;" autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0"></textarea>
			</div>
			<div class="CodeMirror-vscrollbar" cm-not-content="true">
				<div style="min-width: 1px;"></div>
			</div>
			<div class="CodeMirror-hscrollbar" cm-not-content="true">
				<div style="height: 100%; min-height: 1px;"></div>
			</div>
			<div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div>
			<div class="CodeMirror-gutter-filler" cm-not-content="true"></div>
			<div class="CodeMirror-scroll" tabindex="-1">
				<div class="CodeMirror-sizer" style="margin-left: 0px;">
					<div style="position: relative;">
						<div class="CodeMirror-lines" role="presentation">
							<div style="position: relative; outline: none;" role="presentation">
								<pre style="height: 0px; overflow: visible;" class="CodeMirror-placeholder">Champ md</pre>
								<div class="CodeMirror-measure">
									<pre><span>xxxxxxxxxx</span></pre>
								</div>
								<div class="CodeMirror-measure"></div>
								<div style="position: relative; z-index: 1;"></div>
								<div class="CodeMirror-cursors"></div>
								<div class="CodeMirror-code" role="presentation"></div>
							</div>
						</div>
					</div>
				</div>
				<div style="position: absolute; height: 30px; width: 1px;"></div>
				<div class="CodeMirror-gutters" style="display: none;"></div>
			</div>
		</div>
		<div class="editor-preview-side"></div>
	</div>
</div>
`;
